pallets/pools/src/tranche.rs

```diff
@@ -106,62 +130,425 @@ pub struct Tranche<Balance, Rate, Weight> {
+		for tranche in self.tranches.iter().rev() {
+			let r = f(tranche)?;
+			res.push(r)
+		}
+		Ok(res)
+	}
+
+	pub fn combine_mut_residual_top<R, F>(&mut self, mut f: F) -> Result<Vec<R>, DispatchError>
+	where
+		F: FnMut(&mut Tranche<Balance, Rate, Weight, CurrencyId>) -> Result<R, DispatchError>,
+	{
+		let mut res = Vec::with_capacity(self.tranches.len());
+		for tranche in self.tranches.iter_mut().rev() {
+			let r = f(tranche)?;
+			res.push(r)
+		}
+		Ok(res)
+	}
+
+	pub fn combine_with_residual_top<R, I, W, F>(
+		&self,
+		with: I,
+		mut f: F,
+	) -> Result<Vec<R>, DispatchError>
+	where
+		F: FnMut(&Tranche<Balance, Rate, Weight, CurrencyId>, W) -> Result<R, DispatchError>,
+		I: IntoIterator<Item = W>,
+	{
+		let mut res = Vec::with_capacity(self.tranches.len());
+		// TODO: Would be nice to error out when with is larger than tranches...
+		let iter = self.tranches.iter().rev().zip(with.into_iter());
+
+		for (tranche, w) in iter {
+			let r = f(tranche, w)?;
+			res.push(r);
+		}
+
+		Ok(res)
+	}
+
+	pub fn combine_with_mut_residual_top<R, W, I, F>(
+		&mut self,
+		with: I,
+		mut f: F,
+	) -> Result<Vec<R>, DispatchError>
+	where
+		F: FnMut(&mut Tranche<Balance, Rate, Weight, CurrencyId>, W) -> Result<R, DispatchError>,
+		I: IntoIterator<Item = W>,
+	{
+		let mut res = Vec::with_capacity(self.tranches.len());
+		// TODO: Would be nice to error out when with is larger than tranches...
+		let iter = self.tranches.iter_mut().rev().zip(with.into_iter());
+
+		for (tranche, w) in iter {
+			let r = f(tranche, w)?;
+			res.push(r);
+		}
+
+		Ok(res)
+	}
+
+	pub fn orders_as_currency<BalanceRatio>(
+		&self,
+		prices: &Vec<BalanceRatio>,
+	) -> Result<Vec<(Balance, Balance)>, DispatchError>
+	where
+		BalanceRatio: FixedPointNumber<Inner = Balance>,
+	{
+		self.combine_with_non_residual_top(prices.iter(), |tranche, price| {
+			tranche.order_as_currency(price)
+		})
+	}
+
+	pub fn calculate_prices<BalanceRatio, Tokens, AccountId>(
+		&mut self,
+		total_assets: Balance,
+		now: Moment,
+	) -> Result<Vec<BalanceRatio>, DispatchError>
+	where
+		BalanceRatio: FixedPointNumber<Inner = Balance>,
+		Tokens: Inspect<AccountId, AssetId = CurrencyId, Balance = Balance>,
+	{
+		let mut remaining_assets = total_assets;
+		let pool_is_zero = total_assets == Zero::zero();
+
+		// we are gonna reverse the order
+		// such that prices are calculated from most senior to junior
+		// there by all the remaining assets are given to the most junior tranche
+		let mut prices = self.combine_mut_residual_top(|tranche| {
+			let total_issuance = Tokens::total_issuance(tranche.currency);
+
+			if pool_is_zero || total_issuance == Zero::zero() {
+				Ok(One::one())
+			} else if tranche.tranche_type == TrancheType::Residual {
+				BalanceRatio::checked_from_rational(remaining_assets, total_issuance)
+					.ok_or(ArithmeticError::Overflow.into())
+			} else {
+				tranche.accrue(now)?;
+				let tranche_value = tranche.balance()?;
```

Comment:
We should probably be consistent in balance vs value naming. I'd say `value = balance * price`, so in this case `tranche.balance()` is just the `tranche_balance`.